src/bufr/DataObject.h

```diff
@@ -0,0 +1,368 @@
+        /// \param compressionLevel The GZip compression level to use, must be 0-9
+        ioda::Variable createVariable(ioda::ObsGroup& obsGroup,
+                                      const std::string& name,
+                                      const std::vector<ioda::Variable>& dimensions,
+                                      const std::vector<ioda::Dimensions_t>& chunks,
+                                      int compressionLevel) const final
+        {
+            auto params = makeCreationParams(chunks, compressionLevel);
+            auto var = obsGroup.vars.createWithScales<T>(name, dimensions, params);
+            var.write(data_);
+            return var;
+        };
+
+        /// \brief Print data to stdout for debug purposes.
+        void print(std::ostream &out) const final
+        {
+            out << "DataObject " << fieldName_ << ":";
+            for (auto val = data_.cbegin(); val != data_.cend(); ++val)
+            {
+                if (val != data_.cbegin()) out << ", ";
+                out << *val;
+            }
+
+            out << std::endl;
+        };
+
+        /// \brief Get the raw data.
+        std::vector<T> getRawData() const { return data_; }
+
+        /// \brief Set the raw data.
+        void setRawData(std::vector<T> data) { data_ = data; }
+
+        /// \brief Get data associated with a given location.
+        /// \param location The location to get data for.
+        /// \return The data at the given location.
+        T get(const Location& loc) const
+        {
+            size_t dim_prod = 1;
+            for (int dim_idx = dims_.size(); dim_idx > static_cast<int>(loc.size()); --dim_idx)
+            {
+                dim_prod *= dims_[dim_idx];
+            }
+
+            // Compute the index into the data array
+            size_t index = 0;
+            for (int dim_idx = loc.size() - 1; dim_idx >= 0; --dim_idx)
+            {
+                index += dim_prod*loc[dim_idx];
+                dim_prod *= dims_[dim_idx];
+            }
+
+            return data_[index];
+        };
+
+        /// \brief Get the size of the data.
+        /// \return The size of the data.
+        size_t size() const { return data_.size(); }
+
+        /// \brief Get the data at the location as an integer.
+        /// \return Integer data.
+        int getAsInt(const Location& loc) const final { return _getAsInt(loc); }
+
+        /// \brief Get the data at the location as a float.
+        /// \return Float data.
+        float getAsFloat(const Location& loc) const final { return _getAsFloat(loc); }
+
+        /// \brief Get the data at the location as a string.
+        /// \return String data.
+        std::string getAsString(const Location& loc) const final { return _getAsString(loc); }
+
+        /// \brief Get the data at the location as a float.
+        /// \return Float data.
+        float getAsFloat(size_t idx) const final { return _getAsFloat(idx); }
```

Comment:
Index vs location? Why is there a second float one?

Reply:
That's a really good question. 

So data object is able to store data of any dimension. It stores all data internally in a 1d vector.

`getAsFloat(const Location& loc)`: The Location is a vector<int> which points to the coordinate of the data point. So if I have 2 dimensional data the location {2, 4} would get me the value at that coordinate. This version of `getAsFloat` computes the correct offset for the data point (inside that 1d vector) based off the coordinate that is given and what it knows about the data's dimensional properties (size of each dimension).

`getAsFloat(size_t idx)`: This function basically allows you to get a data point based on the idx in the 1d array (it allows for direct access). In general it is not as safe, but computationally a lot cheaper than the previous example. I added it to get date time data making the assumption that this data is always 1d... In retrospect this was probably a stupid thing to do, however. We might want to refactor some of this code later and remove this function.

Reply:
I updated some of the doc strings to make this a little clearer